### PR TITLE
SslStream WriteAsync

### PR DESCRIPTION
**Pull request by benaadams**
Implement WriteAsync for SslStream

Resolves https://github.com/dotnet/corefx/issues/14698

### CODE REVIEW

**Comment by benaadams** [comment]
@dotnet-bot test outerloop CentOS7.1 Debug
@dotnet-bot test outerloop CentOS7.1 Release
@dotnet-bot test outerloop OSX Debug
@dotnet-bot test outerloop OSX Release
@dotnet-bot test outerloop Ubuntu14.04 Debug
@dotnet-bot test outerloop Ubuntu14.04 Release
@dotnet-bot test outerloop Windows 7 Debug
@dotnet-bot test outerloop Windows 7 Release
@dotnet-bot test outerloop Windows_NT Debug
@dotnet-bot test outerloop Windows_NT Release

**Comment by benaadams** [comment]
@CIPop @Priya91 @cesarbs PTAL

Don't think the outerloop failures are related to this change

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 1]
```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
...
+        }
+
+        private async Task WriteAsyncImpl(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var semaphore = EnsureAsyncActiveWriteSemaphoreInitialized();
+            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            InitiateWrite(buffer, offset, count, null);
+
+            bool failed = false;
```
Comment (cesarbs) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: nit: `var`?

**Comment by cesarbs** [comment]
Checking if this fixes the hang in Kestrel.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 2]
```diff
@@ -492,6 +557,81 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
             }
         }
 
+        private async Task StartWritingAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // We loop to this method from the callback.
```
Comment (CIPop) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Nit: Most of this is copy/paste from `StartWriting`. It would be nice to refactor and reuse (extract method) the common wrapper code and only keep different code in two separate functions. I saw only 3 parts that are different (maybe some sort of template method pattern would work here?)

**Comment by CIPop** [comment]
Thanks @benaadams ! 

> Resolves #14698

Aren't ReadAsync and NegotiateStream fixes required before we can close on the general issue?

File: src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs [hunk 3]
```diff
@@ -260,6 +260,31 @@ public void SslStream_StreamToStream_Write_ReadByte_Success()
...
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var clientSslStream = new SslStream(clientStream, false, AllowAnyServerCertificate))
+            using (var serverSslStream = new SslStream(serverStream))
+            {
+                bool result = DoHandshake(clientSslStream, serverSslStream);
+                Assert.True(result, "Handshake completed.");
+
+                for (int i = 0; i < 3; i++)
+                {
+                    await clientSslStream.WriteAsync(_sampleMsg, 0, _sampleMsg.Length).ConfigureAwait(false);
```
Comment (benaadams) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs]: Overzealous; normally forget to added them 😄

**Comment by cesarbs** [comment]
Kestrel HTTPS works again with this fix 🎉 

Just to be sure I tested this correctly, please review my steps (first time doing this):

1) I cloned corefx
2) Fetched @benaadams's PR
3) sync.cmd
4) clean.cmd
5) build.cmd

After that, I pointed Kestrel to my local CoreFX packages by adding

```
<add key="Local CoreFX" value="D:\src\dotnet\corefx\bin.localpkg" />
```

to NuGet.config.

Was that the right place to point to?

**Comment by Priya91** [comment]
> Was that the right place to point to?

yes, that's correct.

File: src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs [hunk 4]
```diff
@@ -260,6 +260,31 @@ public void SslStream_StreamToStream_Write_ReadByte_Success()
 
         [OuterLoop] // TODO: Issue #11345
         [Fact]
+        public async Task SslStream_StreamToStream_WriteAsync_ReadByte_Success()
+        {
+            VirtualNetwork network = new VirtualNetwork();
```
Comment (cesarbs) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs]: Would be interesting to have a test for the specific case that prompted this change - a task blocked on `ReadAsync` waiting for that, while the test calls `WriteAsync`. The inner stream should be guaranteed to not implement begin/end methods (what if `VirtualNetworkStream` is changed in the future?).

**Comment by CIPop** [comment]
Summary of Outerloop failures: I agree with @benaadams that none are related to this change.

* Centos DBG: failure connecting to remote server (already under investigation, tracked by #14519):
```
System.Net.Http.Functional.Tests.HttpClientHandler_SslProtocols_Test.GetAsync_SupportedSSLVersion_Succeeds(name: \"TLSv1.1\", url: \"https://www.ssllabs.com:10302/\") [FAIL]
21:57:46         System.Net.Http.HttpRequestException : An error occurred while sending the request.
```

* Centors RET: Read operation timeout in VirtualNetwork for NegotiateStream. I don't see any relation to the current changes although it could be related to the bug itself. VirtualNetwork is sensitive to timing and will time-out on a busy system. Tracking with #14519.

```
MESSAGE:
System.AggregateException : One or more errors occurred. (The read operation failed, see inner exception.) (VirtualNetwork: Timeout reading the next frame.)\n---- System.IO.IOException : The read operation failed, see inner exception.\n-------- System.TimeoutException : VirtualNetwork: Timeout reading the next frame.\n---- System.TimeoutException : VirtualNetwork: Timeout reading the next frame.
+++++++++++++++++++
STACK TRACE:
at System.ThrowHelper.ThrowAggregateException(List`1 exceptions) at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken) at System.Net.Security.Tests.NegotiateStreamTest.NegotiateStream_EchoServer_ClientWriteRead_Successive_Async_Success() in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/System.Net.Security/tests/FunctionalTests/NegotiateStreamTestForUnix.cs:line 375 ----- Inner Stack Trace #1 (System.IO.IOException) ----- at System.Net.Security.NegotiateStream.EndRead(IAsyncResult asyncResult) at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.Net.Security.Tests.NegotiateStreamTest.<ReadAllAsync>d__17.MoveNext() in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/System.Net.Security/tests/FunctionalTests/NegotiateStreamTestForUnix.cs:line 385 ----- Inner Stack Trace ----- at System.Net.Test.Common.VirtualNetwork.ReadFrame(Boolean server, Byte[]& buffer) in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/Common/tests/System/Net/VirtualNetwork/VirtualNetwork.cs:line 39 at System.Net.Test.Common.VirtualNetworkStream.Read(Byte[] buffer, Int32 offset, Int32 count) in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/Common/tests/System/Net/VirtualNetwork/VirtualNetworkStream.cs:line 119 at System.IO.Stream.<>c.<BeginReadInternal>b__38_0(Object <arg>) at System.Threading.Tasks.Task`1.InnerInvoke() at System.Threading.Tasks.Task.Execute() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.IO.Stream.EndRead(IAsyncResult asyncResult) at System.Net.FixedSizeReader.ReadCallback(IAsyncResult transportResult) ----- Inner Stack Trace #2 (System.TimeoutException) ----- at System.Net.Test.Common.VirtualNetwork.ReadFrame(Boolean server, Byte[]& buffer) in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/Common/tests/System/Net/VirtualNetwork/VirtualNetwork.cs:line 39 at System.Net.Test.Common.VirtualNetworkStream.Read(Byte[] buffer, Int32 offset, Int32 count) in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/Common/tests/System/Net/VirtualNetwork/VirtualNetworkStream.cs:line 119 at System.Net.Security.Tests.UnixGssFakeStreamFramer.ReadDataFrame() in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/System.Net.Security/tests/FunctionalTests/UnixGssFakeStreamFramer.cs:line 82 at System.Net.Security.Tests.UnixGssFakeNegotiateStream.MessageLoop(Object state) in /mnt/resource/j/workspace/dotnet_corefx/master/outerloop_centos7.1_release_prtest/src/System.Net.Security/tests/FunctionalTests/UnixGssFakeNegotiateStream.cs:line 75 at System.Threading.Tasks.Task.Execute()
```

* Ubuntu DBG, RET: timeout (System.Net.Security.FunctionalTests/UnitTests completed)

* Win 7 Certificate Validation issue. (Known issue: #13637)

```
System.Net.Security.Tests.SslStreamSystemDefaultTest.SslStream_DefaultTlsConfigurationSync_Ok [FAIL]
21:54:59         System.Security.Authentication.AuthenticationException : The remote certificate is invalid according to the validation procedure.
```

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 2 above)
Comment (benaadams) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: The main bit that is essentially different is a call to `WriteAsync` vs `Write` etc, and returning a Task. However, trying to extract the methods made it messy with parameters. Went for a quick fix, rather than something I'm proud of 😢 It can be made much better; though needs to support three forms of Write...

File: src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs (hunk 4 above)
Comment (benaadams) [inline src/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs]: Would likely have the issue again, if the inner stream didn't implement WriteAsync or ReadAsync

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 5]
```diff
@@ -30,6 +31,8 @@ internal class SslStreamInternal
         private AsyncProtocolRequest _readProtocolRequest; // cached, reusable AsyncProtocolRequest used for read operations
         private AsyncProtocolRequest _writeProtocolRequest; // cached, reusable AsyncProtocolRequest used for write operations
 
+        private SemaphoreSlim _asyncWriteActiveSemaphore;
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Why is this needed?  The existing BeginWrite doesn't serialize on such a semaphore, does it?

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 6]
```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
...
+            InitiateWrite(buffer, offset, count, null);
+
+            bool failed = false;
+            try
+            {
+                await StartWritingAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+
+                if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "WriteAsyncImpl"));
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Is this correct?  invalidendcall isn't relevant for WriteAsync, and WriteAsyncImpl isn't publically exposed, so that name doesn't mean anything to a caller.  Is this actually dead code, or if not, in what situations is this possible?

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Specifically, the existing code uses _nestedWrite to throw an exception if multiple writes are issued concurrently; in contrast, this change is making it so that it's allowed to issue multiple concurrent write requests, and they just end up getting serialized.  If that's actually a desired change and deviation from the .NET Framework, it should be made as its own separate and deliberate change, and should likely apply to BeginWrite as well.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 7]
```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
...
+                if (e is IOException)
+                {
+                    throw;
+                }
+
+                throw new IOException(SR.net_io_write, e);
+            }
+            finally
+            {
+                semaphore.Release();
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Once the semaphore is removed, then this finally block can just go away, the failed bool can go away, and _nestedWrite can be set to 0 in the catch block.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Then at that point, it seems this whole WriteAsyncImpl method could go away, just adding a catch block to StartWritingAsync and avoiding the need for an extra level of async method, which will result in an extra set of allocations if StartWritingAsync method yields.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Then at that point, it seems this whole WriteAsyncImpl method could go away, just adding a catch block to StartWritingAsync and avoiding the need for an extra level of async method, which will result in an extra set of allocations if StartWritingAsync method yields.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 8]
```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
...
+                        ? Task.FromCanceled(cancellationToken)
+                        : WriteAsyncImpl(buffer, offset, count, cancellationToken);
+        }
+
+        private async Task WriteAsyncImpl(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var semaphore = EnsureAsyncActiveWriteSemaphoreInitialized();
+            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            InitiateWrite(buffer, offset, count, null);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Previously wouldn't any exceptions from InitiateWrite (e.g. another operation in progress) have emerged synchronously from the call to WriteAsync?  Now they're going to be stored into the returned task.  Following my other comments, this InitiateWrite call can just be moved to the WriteAsync caller.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 9]
```diff
@@ -384,6 +433,22 @@ private void ProcessWrite(byte[] buffer, int offset, int count, LazyAsyncResult
...
+
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "Write", "write"));
+            }
+
+            // If this is an async operation, get the AsyncProtocolRequest to use.
+            // We do this only after we verify we're the sole write operation in flight.
+            AsyncProtocolRequest asyncRequest = GetOrCreateProtocolRequest(ref _writeProtocolRequest, asyncResult);
+            return asyncRequest;
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Nit: can just be ```return GetOrCreateProtocolRequest(ref _writeProtocolRequest, asyncResult);```

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: > It can be made much better

Is there an issue tracking that?

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 10]
```diff
@@ -492,6 +557,81 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
...
+                        throw new IOException(SR.net_io_encrypt, message.GetException());
+                    }
+
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    {
+                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWritingAsync Got Encrypted Buffer",
+                            this.GetHashCode(), encryptedBytes, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
+                    }
+
+                    await _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes, cancellationToken);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: This is missing a ConfigureAwait(false)

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: > We loop to this method from the callback. 

What "callback"?

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 9 above)
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Actually, why is this code even part of InitiateWrite?  InitiateWrite is used in two places: from ProcessWrite (part of Begin/EndWrite) and from WriteAsyncImpl... but WriteAsyncImpl just ignores this return value.  It would be better to just make InitiateWrite return void and leave this GetOrCreateProtocolRequest code in ProcessWrite, where it's actually needed.

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 11]
```diff
@@ -492,6 +557,81 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
...
+                if (_pinnableOutputBufferInUse == null)
+                {
+                    if (_pinnableOutputBuffer == null)
+                    {
+                        _pinnableOutputBuffer = s_PinnableWriteBufferCache.AllocateBuffer();
+                    }
+
+                    _pinnableOutputBufferInUse = buffer;
+                    outBuffer = _pinnableOutputBuffer;
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: @cipop, do we really need all of this event source stuff?  The comment on PinnableBufferCacheEventSource suggests that it's just for examining the effectiveness of the PinnableBufferCache... is anyone actually paying attention to it, or is it just unnecessary noise and cost?

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs [hunk 12]
```diff
@@ -345,17 +404,7 @@ private AsyncProtocolRequest GetOrCreateProtocolRequest(ref AsyncProtocolRequest
...
-
-            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
-            {
-                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "Write", "write"));
-            }
-
-            // If this is an async operation, get the AsyncProtocolRequest to use.
-            // We do this only after we verify we're the sole write operation in flight.
-            AsyncProtocolRequest asyncRequest = GetOrCreateProtocolRequest(ref _writeProtocolRequest, asyncResult);
+            var asyncRequest = InitiateWrite(buffer, offset, count, asyncResult);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: Nit: per my other comments, InitiateWrite should really return void, but if for some reason it stays returning an AsyncProtocolRequest, then the ```var``` should be changed to ```AsyncProtocolRequest```.

**Comment by stephentoub** [comment]
I'll submit a PR to fix some of the issues I called out.

**Comment by stephentoub** [comment]
> I'll submit a PR to fix some of the issues I called out.

Actually, looking at it closer, I'm not clear on how other aspects of this change are valid.  I see the existing code doing checks for whether a handshake is currently in progress, and all of that was removed for the new WriteAsync code paths... @cipop, @benaadams, why is that valid?

**Comment by benaadams** [comment]
Might not be, I implemented it by following the stack trace of what it was currently doing in the blocked state
```
Stream.WriteAsync
 -> Stream.BeginEndWriteAsync
 -> TaskFactory.FromAsync
 -> Stream.BeginEndWriteAsync...
 -> SslStream.BeginWrite
 -> SslStreamInternal.ProcessWrite
 -> SslStreamInternal.StartWriting
 -> Stream.BeginWrite
 -> Stream.BeginWriteInternal
```

File: src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs (hunk 11 above)
Comment (CIPop) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: @brianrob @vancem own this portion in NetFX. Since it needs to be enabled I don't see the noise and the cost of the single if should not be significant to justify a change.

Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs]: I wasn't questioning the cost of one if, but rather all of the code in this method related to this. If no one's looking at it, it's wasteful, if for nothing else than code complexity.
